components/push/src/lib.rs

```diff
@@ -4,6 +4,181 @@
+//!
+//! This subscription info is distributed to other services that want to send push messages to
+//! the application.
+//!
+//! The HTTP endpoint is provided by Mozilla's [autopush service](https://autopush.readthedocs.io/en/latest/),
+//! and we use the [rust-ece](https://github.com/mozilla/rust-ece) to manage encryption with the cryptographic keys.
+//!
+//! Here's a helpful diagram of how the *subscription* flow works at a high level across the moving parts:
+//! ![A Sequence diagram showing how the different parts of push interact](https://mozilla.github.io/application-services/book/diagrams/Push-Component-Subscription-flow.png "Sequence diagram")
+//!
+//! ### AutoPush Bridging
+//!
+//! Our target consumer platforms each have their own proprietary push-notification infrastructure,
+//! such as [Firebase Cloud Messaging](https://firebase.google.com/docs/cloud-messaging) for Android
+//! and the [Apple Push Notification Service](https://developer.apple.com/notifications/) for iOS.
+//! Mozilla's [autopush service](https://autopush.readthedocs.io/en/latest/) provides a bridge between
+//! these different mechanisms and the WebPush standard so that they can be used with a consistent
+//! interface.
+//!
+//! This component acts a client of the [Push Service Bridge HTTP Interface](https://autopush.readthedocs.io/en/latest/http.html#push-service-bridge-http-interface).
+//!
+//! We assume two things about the consuming application:
+//! * It has registered with the autopush service and received a unique `app_id` identifying this registration.
+//! * It has registred with whatever platform-specific notification infrastructure is appropriate, and is
+//!   able to obtain a `token` corresponding to its native push notification state.
+//!
+//! On first use, this component will register itself as an *application instance* with the autopush service, providing the `app_id` and `token` and receiving a unique `uaid` ("user-agent id") to identify its
+//! connection to the server.
+//!
+//! As the application adds or removes subscriptions using the API of this component, it will:
+//! * Manage a local database of subscriptions and the corresponding cryptographic material.
+//! * Make corresponding HTTP API calls to update the state associated with its `uaid` on the autopush server.
+//!
+//! Periodically, the application should call a special `verify_connection` method to check whether
+//! the state on the autopush server matches the local state and take any corrective action if it
+//! differs.
+//!
+//! For local development and debugging, it is possible to run a local instance of the autopush
+//! bridge service; see [this google doc](https://docs.google.com/document/d/18L_g2hIj_1mncF978A_SHXN4udDQLut5P_ZHYZEwGP8) for details.
+//!
+//! ## API
+//!
+//! ## Initialization
+//!
+//! Calls are handled by the `PushManager`, which provides a handle for future calls.
+//!
+//! example:
+//! ```kotlin
```

Comment:
Heh, interesting...I think it's lightly weird to have Kotlin example code in a `.rs` source file, but also, this *does* seem inline with the discussion of how might consume these docs, so I'm happy to run with it!